### PR TITLE
Add benchamark

### PR DESCRIPTION
**Pull request by PikachuEXE**
My result:
```
→ ruby benchmark/main.rb
-- create_table(:blogs, {:force=>true})
   -> 0.0035s
-- create_table(:posts, {:force=>true})
   -> 0.0004s
-- create_table(:users, {:force=>true})
   -> 0.0003s
-- create_table(:addresses, {:force=>true})
   -> 0.0003s
-- create_table(:groups, {:force=>true})
   -> 0.0003s
-- create_table(:tags, {:force=>true})
   -> 0.0003s
-- create_table(:post_tags, {:force=>true})
   -> 0.0003s
                           user     system      total        real
AR eager loading:      0.017352   0.000509   0.017861 (  0.017874)
AR with goldiloader:   0.032490   0.000850   0.033340 (  0.033918)
```

### CODE REVIEW

File: benchmark/main.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  user_2 = User.create!
+  Post.create!(author: user_1, blog: blog_1)
+  Post.create!(author: user_2, blog: blog_2)
+end
+
+Benchmark.bm(20) do |x|
+  # Use AR's eager loading
+  x.report("AR eager loading: ") do
+    ::Blog.all.includes(posts: :author).map do |blog|
+      blog.posts.to_a.each do |post|
```
Comment (jturkel) [inline benchmark/main.rb]: I'm not sure it matters but this benchmark will generate less garbage if we drop the `to_a` and call `blog.posts.each`.

**Comment by PikachuEXE** [comment]
Benchmark updated
```
→ ruby benchmark/main.rb
-- create_table(:blogs, {:force=>true})
   -> 0.0088s
-- create_table(:posts, {:force=>true})
   -> 0.0004s
-- create_table(:users, {:force=>true})
   -> 0.0006s
-- create_table(:addresses, {:force=>true})
   -> 0.0003s
-- create_table(:groups, {:force=>true})
   -> 0.0003s
-- create_table(:tags, {:force=>true})
   -> 0.0007s
-- create_table(:post_tags, {:force=>true})
   -> 0.0003s
Warming up --------------------------------------
  AR eager loading:      6.000  i/100ms
AR with goldiloader:
                         3.000  i/100ms
Calculating -------------------------------------
  AR eager loading:      68.463  (± 4.4%) i/s -    342.000  in   5.003666s
AR with goldiloader:
                         36.166  (± 5.5%) i/s -    183.000  in   5.076981s

Comparison:
  AR eager loading: :       68.5 i/s
AR with goldiloader: :       36.2 i/s - 1.89x  slower

```

File: benchmark/db/schema.rb [hunk 2]
```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
+
+ActiveRecord::Schema.define(version: 0) do
```
Comment (jturkel) [inline benchmark/db/schema.rb]: It looks like this was copied from `spec/db/schema.rb`. If we think benchmarks and correctness tests will use the same schemas can we share that file across both types of tests. Alternatively we think the schemas for benchmarks and correctness tests will diverge, can we slim down `benchmark/db/schema.rb` to just the schema elements needed by the current benchmarks.

File: benchmark/main.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        post.author.id
+      end
+    end
+  end
+
+  require "goldiloader"
+
+  # Use goldiloader
+  x.report("AR with goldiloader: ") do
+    ::Blog.all.map do |blog|
```
Comment (jturkel) [inline benchmark/main.rb]: This `map` should be an `each` to be consistent with the "AR eager loading" test.

File: benchmark/main.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+$:.push File.expand_path("lib", __dir__)
```
Comment (jturkel) [inline benchmark/main.rb]: Should this be `File.expand_path("../lib", __dir__)` so it points at the project root directory.

**Comment by PikachuEXE** [comment]
All issues fixed including rubocop

**Comment by PikachuEXE** [comment]
Not sure if there is a way to run benchmark in CI...
Or should we?

**Comment by jturkel** [comment]
I'd love to have benchmarks running in CI that can automatically flag performance regressions but I haven't seen this work well in oversubscribed CI environments with lots of noisy neighbors. I'm curious if others have seen this work well in other projects.

**Pull request merged**